include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,512 @@
+        : Node(c) {
+      assert(c == B || c == BB);
+    }
+    virtual bool isLeaf() const override { return true; }
+    virtual ~Leaf(){};
+  };
+
+  struct InternalNode : public Node {
+    InternalNode(
+        Color c, std::shared_ptr<const Node> const &lft, T key, V val,
+        std::shared_ptr<const Node> const &rgt)
+        : Node(c)
+        , _lft(lft)
+        , _key(key)
+        , _val(val)
+        , _rgt(rgt) {
+      this->_s = 1 + _lft->_s + _rgt->_s;
+    }
+    std::shared_ptr<const Node> _lft;
+    T _key;
+    V _val;
+    std::shared_ptr<const Node> _rgt;
+
+    virtual bool isLeaf() const override { return false; }
+    virtual ~InternalNode(){};
+  };
+
+  explicit RBTree(std::shared_ptr<const Node> const &node)
+      : _root(node) { }
+
+  Color rootColor() const {
+    assert(!isEmpty());
+    return _root->_c;
+  }
+
+  Color leafColor() const {
+    assert(isEmpty());
+    return _root->_c;
+  }
+
+public:
+  RBTree()
+      : _root(std::make_shared<const Leaf>(B)) { }
+
+  RBTree(Color c)
+      : _root(std::make_shared<const Leaf>(c)) { }
+
+  RBTree(Color c, RBTree const &lft, T key, V val, RBTree const &rgt)
+      : _root(std::make_shared<const InternalNode>(
+          c, lft._root, key, val, rgt._root)) {
+    assert(lft.isEmpty() || lft.root() < key);
+    assert(rgt.isEmpty() || key < rgt.root());
+  }
+
+  template <class I>
+  RBTree(I b, I e) {
+    RBTree t(B);
+    for_each(b, e, [&t](std::pair<T, V> const &p) {
+      t = t.inserted(p.first, p.second);
+    });
+    _root = t._root;
+  }
+
+  bool isEmpty() const { return _root->isLeaf(); }
+
+  T root() const {
+    assert(!isEmpty());
+    const InternalNode *r = static_cast<const InternalNode *>(_root.get());
+    return r->_key;
+  }
+
+  V rootVal() const {
+    assert(!isEmpty());
+    const InternalNode *r = static_cast<const InternalNode *>(_root.get());
+    return r->_val;
+  }
+
+  RBTree left() const {
+    assert(!isEmpty());
+    const InternalNode *r = static_cast<const InternalNode *>(_root.get());
+    return RBTree(r->_lft);
+  }
+
+  RBTree right() const {
+    assert(!isEmpty());
+    const InternalNode *r = static_cast<const InternalNode *>(_root.get());
+    return RBTree(r->_rgt);
+  }
+
+  size_t size() const { return _root->_s; }
+
+  bool member(T x) const {
+    if (isEmpty())
+      return false;
+    T y = root();
+    if (x < y)
+      return left().member(x);
+    else if (y < x)
+      return right().member(x);
+    else
+      return true;
+  }
+
+  V lookup(T x) const {
+    if (isEmpty())
+      CONSTRUCT_MSG_AND_THROW("Key not found for map lookup");
+    T y = root();
+    if (x < y)
+      return left().lookup(x);
+    else if (y < x)
+      return right().lookup(x);
+    else
+      return rootVal();
+  }
+
+  RBTree inserted(T x, V v) const { return ins(x, v).blacken(); }
+
+  RBTree deleted(T x) const { return redden().del(x); }
+
+  // 1. No red node has a red child.
+  void assert1() const {
```

Comment:
This could have a better name if we're exposing it as part of a public interface.